### ISSUE TITLE
[release][core] chaos_many_actors is failing due to ray.wait regression

### ISSUE THREAD

**Issue by xwjiang2010** [ray-project/ray#30694]
https://console.anyscale-staging.com/o/anyscale-internal/projects/prj_qC3ZfndQWYYjx2cz8KWGNUL4/clusters/ses_j43zvvuj4jtf2jtrzfislqye?command-history-section=command_history

**Comment by iycheng** [comment]
Got some useful logs.
```
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(DBActor pid=4559) [2022-11-28 10:32:45,917 E 4559 4559] actor_scheduling_queue.cc:85: client skipping requests 0 to 0
2022-11-28 10:32:46,437 WARNING worker.py:1851 -- WARNING: 224 PYTHON worker processes have been started on node: b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf with address: 172.31.206.202. This could be a result of using a large number of actors, or due to tasks blocked in ray.get() calls (see https://github.com/ray-project/ray/issues/3644 for some discussion of workarounds).
```

**Comment by iycheng** [comment]
I feel we should set `max_restarts=-1, max_task_retries=-1` for DBActor too. Still confirming with people.

**Comment by iycheng** [comment]
Can't figure out what's happening..... return the test

**Comment by iycheng** [comment]
The rerun finished successfully.

**Comment by iycheng** [comment]
```
Chaos test:   3%|▎         | 2777/86400 [01:03<28:41, 48.59it/s](scheduler +44s) Tip: use `ray status` to view detailed cluster status. To disable these messages, set RAY_SCHEDULER_EVENTS=0.
(scheduler +44s) Resized to 144 CPUs.
Traceback (most recent call last):
  File "chaos_test/test_chaos_basic.py", line 235, in <module>
    main()
  File "chaos_test/test_chaos_basic.py", line 188, in main
    workload(total_num_cpus, args.smoke)
  File "chaos_test/test_chaos_basic.py", line 106, in run_actor_workload
    pb.fetch_until_complete(results)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/data/_internal/progress_bar.py", line 75, in fetch_until_complete
    for ref, result in zip(done, ray.get(done)):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/worker.py", line 2309, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError: ray::ReportActor.add() (pid=216, ip=172.31.72.49, repr=<test_chaos_basic.ReportActor object at 0x7fdcb476e050>)
  File "chaos_test/test_chaos_basic.py", line 82, in add
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(DBActor pid=4559) [2022-11-28 10:32:45,917 E 4559 4559] actor_scheduling_queue.cc:85: client skipping requests 0 to 0
2022-11-28 10:32:46,437 WARNING worker.py:1851 -- WARNING: 224 PYTHON worker processes have been started on node: b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf with address: 172.31.206.202. This could be a result of using a large number of actors, or due to tasks blocked in ray.get() calls (see https://github.com/ray-project/ray/issues/3644 for some discussion of workarounds).

```

**Comment by iycheng** [comment]
```
[2022-11-28 10:31:47,409 I 216 412] actor_manager.cc:214: received notification on actor, state: PENDING_CREATION, actor_id: 742497be0149fea563a74cee03000000, ip address: , port: 0, worker_id: NIL_ID, raylet_id: b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf, num_restarts: 0, death context type=CONTEXT_NOT_SET
```

**Comment by iycheng** [comment]
```
[2022-11-28 10:31:42,741 I 291 291] (gcs_server) gcs_actor_manager.cc:247: Registering actor, job id = 03000000, actor id = 742497be0149fea563a74cee03000000
[2022-11-28 10:31:42,743 I 291 291] (gcs_server) gcs_actor_manager.cc:253: Registered actor, job id = 03000000, actor id = 742497be0149fea563a74cee03000000
[2022-11-28 10:31:42,746 I 291 291] (gcs_server) gcs_actor_manager.cc:272: Creating actor, job id = 03000000, actor id = 742497be0149fea563a74cee03000000
[2022-11-28 10:31:42,746 I 291 291] (gcs_server) gcs_actor_scheduler.cc:308: Start leasing worker from node b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf for actor 742497be0149fea563a74cee03000000, job id = 03000000
[2022-11-28 10:32:40,151 I 291 291] (gcs_server) gcs_actor_scheduler.cc:627: Finished leasing worker from b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf for actor 742497be0149fea563a74cee03000000, job id = 03000000
[2022-11-28 10:32:40,152 I 291 291] (gcs_server) gcs_actor_scheduler.cc:440: Start creating actor 742497be0149fea563a74cee03000000 on worker b183cf22ee1cc85b294e0ed82848e9231a3e364efe7caf9faf6c079d at node b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf, job id = 03000000
[2022-11-28 10:32:40,156 I 291 291] (gcs_server) gcs_actor_scheduler.cc:479: Succeeded in creating actor 742497be0149fea563a74cee03000000 on worker b183cf22ee1cc85b294e0ed82848e9231a3e364efe7caf9faf6c079d at node b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf, job id = 03000000
[2022-11-28 10:32:40,156 I 291 291] (gcs_server) gcs_actor_manager.cc:1186: Actor created successfully, actor id = 742497be0149fea563a74cee03000000, job id = 03000000
[2022-11-28 10:32:40,156 I 291 291] (gcs_server) gcs_actor_manager.cc:289: Finished creating actor, job id = 03000000, actor id = 742497be0149fea563a74cee03000000
```

**Comment by iycheng** [comment]
```
[2022-11-28 10:31:42,746 I 291 291] (gcs_server) gcs_actor_scheduler.cc:308: Start leasing worker from node b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf for actor 742497be0149fea563a74cee03000000, job id = 03000000
[2022-11-28 10:32:40,151 I 291 291] (gcs_server) gcs_actor_scheduler.cc:627: Finished leasing worker from b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf for actor 742497be0149fea563a74cee03000000, job id = 03000000
```

It seems use a lot of time finish leasing the worker.

**Comment by iycheng** [comment]
So bad... tried 5 times and all passed...

**Comment by xwjiang2010** [comment]
should we drop release blocker then and treat it as flaky test?

**Comment by iycheng** [comment]
> should we drop release blocker then and treat it as flaky test?

@scv119  how do you think?

**Comment by xwjiang2010** [comment]
Timing out on newest try: https://buildkite.com/ray-project/release-tests-branch/builds/1227#0184c6bc-2139-4885-9ddc-63b308239b01

**Comment by iycheng** [comment]
I'm still investigating on this one. I think the difficulty to reproduce on my side make it slow.

**Comment by iycheng** [comment]
logs: [ses_j43zvvuj4jtf2jtrzfislqye.tar.gz](https://github.com/ray-project/ray/files/10127373/ses_j43zvvuj4jtf2jtrzfislqye.tar.gz)

**Comment by iycheng** [comment]
Actor 742497be0149fea563a74cee03000000 timeline:

GCS LOG:
```
 [2022-11-28 10 31 42,741 I 291 291] Registering actor, job id = 03000000, actor id = 742497be0149fea563a74cee03000000
 [2022-11-28 10 31 42,743 I 291 291] Registered actor, job id = 03000000, actor id = 742497be0149fea563a74cee03000000
 [2022-11-28 10 31 42,746 I 291 291] Creating actor, job id = 03000000, actor id = 742497be0149fea563a74cee03000000
 [2022-11-28 10 31 42,746 I 291 291] Start leasing worker from node b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf for actor 742497be0149fea563a74cee03000000, job id = 03000000
 [2022-11-28 10 32 40,151 I 291 291] Finished leasing worker from b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf for actor 742497be0149fea563a74cee03000000, job id = 03000000
 [2022-11-28 10 32 40,152 I 291 291] Start creating actor 742497be0149fea563a74cee03000000 on worker b183cf22ee1cc85b294e0ed82848e9231a3e364efe7caf9faf6c079d at node b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf, job id = 03000000
 [2022-11-28 10 32 40,156 I 291 291] Succeeded in creating actor 742497be0149fea563a74cee03000000 on worker b183cf22ee1cc85b294e0ed82848e9231a3e364efe7caf9faf6c079d at node b277fed6035777a7093591b49a8c71859bb94da55bc4e4dbbbd743cf, job id = 03000000
 [2022-11-28 10 32 40,156 I 291 291] Actor created successfully, actor id = 742497be0149fea563a74cee03000000, job id = 03000000
 [2022-11-28 10 32 40,156 I 291 291] Finished creating actor, job id = 03000000, actor id = 742497be0149fea563a74cee03000000

**Comment by iycheng** [comment]
```
[2022-11-28 10:32:04,734 I 2081 2168] core_worker.cc:326: Will resubmit task after a 0ms delay: Type=ACTOR_TASK, Language=PYTHON, Resources: {}, function_descriptor={type=PythonFunctionDescriptor, module_name=test_chaos_basic, class_name=run_actor_workload.<locals>.ReportActor, function_name=add, function_hash=}, task_id=3d8005035b782ccf721a900fd5a976b3054ea01803000000, task_name=ReportActor.add, job_id=03000000, num_args=2, num_returns=2, depth=0, attempt_number=2, actor_task_spec={actor_id=721a900fd5a976b3054ea01803000000, actor_caller_id=ffffffffffffffffffffffffffffffffffffffff03000000, actor_counter=138}
```

**Comment by iycheng** [comment]
The timeout one is because the scheduling perf is slow:
```
Chaos test:   3%|▎         | 2862/86400 [02:29<1:29:38, 15.53it/s]
```

which previously, it's 30it/s

**Comment by iycheng** [comment]
https://github.com/ray-project/ray/issues/30606

**Comment by scv119** [comment]
```
(DBActor pid=4559) [2022-11-28 10:32:45,917 E 4559 4559] actor_scheduling_queue.cc:85: client skipping requests 0 to 0
```
seems to be the root cause for the failure https://console.anyscale-staging.com/o/anyscale-internal/projects/prj_qC3ZfndQWYYjx2cz8KWGNUL4/clusters/ses_j43zvvuj4jtf2jtrzfislqye?command-history-section=command_history

**Comment by iycheng** [comment]
The offending one is: https://github.com/ray-project/ray/commit/7bccdb777cf3a8d9d3696144c0e97ef3792d6083 @cadedaniel  could you take a look?

This test is testing with node failures so it mean s a lot of objects can be lost.

The commit makes the test run slower and then timeout.

**Comment by ericl** [comment]
Hmm. A bit surprising since test_chaos_basic.py has no use of `ray.wait` anywhere.

**Comment by ericl** [comment]
Oh wait, it does have a ray.wait() buried through the use of `ProgressBar.fetch_until_complete()`--- and this is a ray.wait on a huge number of actor tasks. So this indeed could be uncovering something significant.

**Comment by ericl** [comment]
A quick fix would be to simply use `block_until_complete` instead of `fetch_until_complete` in the test, since we never look at the results anyways, but it's probably a good use of time to understand the root cause of this test failure.

I would recommend time-boxing this and considering moving the change to 2.3 if it turns out to be nontrivial.

**Comment by iycheng** [comment]
I'm running it with block_until_complete to see whether it can mitigate this.

**Comment by iycheng** [comment]
It passed https://console.anyscale.com/o/anyscale-internal/projects/prj_FKRmeV5pA6X72aVscFALNC32/clusters/ses_l1nte53neab2urk21ee4tv4n
with the change.

**Comment by cadedaniel** [comment]
I believe I found the root cause, testing a fix.

**Comment by cadedaniel** [comment]
Okay, update. The fix did not have the impact I expected. I think we need to revert the change https://github.com/ray-project/ray/commit/7bccdb777cf3a8d9d3696144c0e97ef3792d6083 from the release branch, given that it causes a performance regression in the following loop:
```python
def problematic(object_refs):
  remaining = object_refs
  while remaining:
    _, remaining = ray.wait(remaining, fetch_local=True)
```

At 10k object refs, this loop runs in ~10s without the change. It jumps to 30s with the change.

I looked into what's going on; it is spending significant additional time in the `RetryObjectInPlasmaErrors` function; from ~4% of samples to ~55% of samples.

##### Without https://github.com/ray-project/ray/commit/7bccdb777cf3a8d9d3696144c0e97ef3792d6083
Blue: CoreWorker::Wait
Pink: CoreWorkerMemoryStore::GetIfExists
Black: RayObject::IsInPlasmaError
<img width="1195" alt="Screen Shot 2022-12-01 at 7 26 47 PM" src="https://user-images.githubusercontent.com/950914/205255261-1dfbeaab-2350-46ce-b821-a9a25355c1d9.png">


##### With https://github.com/ray-project/ray/commit/7bccdb777cf3a8d9d3696144c0e97ef3792d6083
<img width="1198" alt="Screen Shot 2022-12-01 at 7 20 40 PM" src="https://user-images.githubusercontent.com/950914/205256206-9402aaa0-82eb-4dc6-b89f-698c98d49c41.png">


The increase is attributable to both `RayObject::IsInPlasmaError` and `CoreWorkerMemoryStore::GetIfExists` taking more overall time (each taking half of the new time). In the tests I ran, `plasma_object_ids.size()` was always zero (this detail will be interesting to @ericl), but it still had to scan the entire `objects_` hashset in `memory_store.cc`, and check each non-nullptr one to see if `IsInPlasmaError` was true.

I can make an issue to find a way to fix the performance regression on the master branch next sprint. It might be possible to keep track of the RayObjects which have `IsInPlasmaError==true` in a separate datastructure in `memory_store.cc`, so that ray.wait can prefetch them from plasma without traversing the entire `objects_` hashset again. Will want to think about this more.

**Comment by ericl** [comment]
Good find, and I agree with the proposed actions (1) revert for now, (2) evaluate the performance and fix for 2.3 (we should also add a microbenchmark style test for waiting on lots of objects).

About the findings themselves, I think a couple things are surprising:
1. That GetIfExists is so expensive compared to the ready.find(). Both are hash table lookups, though maybe the GetIfExists is also acquiring a mutex and doing lookups on a much bigger hash table, so it could be slower.
2. That this is alone responsible for 10->30s regression. Could we also be adding a lot of overhead on the raylet side to process the additional wait request ids?

Regarding potential fixes, I think one robust way to fix this may be to limit the number of non-ready obj ids we send to raylet to prefetch. Previously we limited it to 0. After the fix, it was unlimited. But some middle ground limit like "at most 200 prefetch ids" may be best for performance. This makes sense since prefetching too many objects adds too much overhead, however, most of the value is just in prefetching a small number of objects.

**Comment by cadedaniel** [comment]
+1 to adding a microbenchmark test for waiting on lots of objects.

I was also surprised that GetIfExists is that expensive. I wrote a batched version (that only obtains the lock once) but the runtime was about the same. Other candidates there are our hash function for ObjectID, but I checked and it's a pretty good implementation (up-to-date 64bit murmurhash), or we're hitting a case where the flat_hash_set has bad performance with this workload and needs tuning. That was further than I decided to go yesterday but may be involved in the perf issue.

Re: GetIfExists and IsInPlasmaError alone causing the additional 20 seconds. I commented out these calls and the runtime went down to 13 seconds. I only measured once so there's some uncertainty there -- but I remember seeing some vector conversion code in the pyx file that was also showing up in the flamegraphs. So you could be right, but the impact is still secondary to GetIfExists and IsInPlasmaError.

On potential fixes: that's a good idea. I was getting stuck on how to do this in a way that is stateless, since we don't know the order in which object ids will be ready. Isn't it a problem if, for object ids [a, b, c, d], if we send [a, b] to the raylet for prefeteching before they're ready, but actually [c, d] are ready first? I think we'll suffer poor prefetching performance, which is the whole point.

**Comment by iycheng** [comment]
remove release blocker from this ticket. assign to @cadedaniel for further work.

**Comment by ericl** [comment]
> Isn't it a problem if, for object ids [a, b, c, d], if we send [a, b] to the raylet for prefeteching before they're ready, but actually [c, d] are ready first?

The logic is a bit misleading; we only are prefetching "ready" ids in any case. If an object exists in the memory store and IsInPlasmaError(), then it is "ready". Unfortunately the "ready" hash set that is passed into RetryObjectInPlasmaErrors is only a subset of the ready objects (limited in size to `num_returns`).

Now, for stateless vs stateful, it depends on where the bottleneck is. If the bottleneck is core worker side, probably sending up to 200 random ids to raylet for prefetching is fine and removes the bottleneck. If the bottleneck is actually raylet side, it gets trickier and we probably want to define an order of prefetching on the ready ids. Otherwise, if you randomly send 200 ids to the raylet to prefetch, eventually the raylet is prefetching all 40000 since it persistently prefetches in the background.

**Comment by ericl** [comment]
I would probably just try out the limiting to 200 and see if it makes a difference at all first.

**Comment by rickyyx** [comment]
Should we be reverting the change in master as well? Since it seems it will take some time to investigate and validate.

**Comment by cadedaniel** [comment]
I will be focusing on this tomorrow, so probably have a fix in by Thursday. IMO fine to revert on master because it could hide other failures in that test that will be a pain to bisect.

**Comment by rkooo567** [comment]
It is reverted iiuc

**Comment by rkooo567** [comment]
I found this is still failing although we reverted this in the master. I created another issue to keep track of it

**Comment by cadedaniel** [comment]
Picking this up again, blocked on https://github.com/ray-project/ray/issues/31760

**Comment by scv119** [comment]
since the offending PR is reverted so we assume it's no longer release blocking. cc @ericl ?

**Comment by rkooo567** [comment]
Since we reverted this, we can just close the issue? Please reopen it if I am wrong
